modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/CheckCatalogVersionOnAppendEntries.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.ignite.raft.jraft.util.Marshaller;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An {@link AppendEntriesRequestInterceptor} that rejects requests (by returning EBUSY error code) if any of the
+ * incoming commands requires catalog version that is not available locally yet.
+ */
+public class CheckCatalogVersionOnAppendEntries implements AppendEntriesRequestInterceptor {
+    private static final IgniteLogger LOG = Loggers.forClass(CheckCatalogVersionOnAppendEntries.class);
+
+    private static final int NO_VERSION_REQUIREMENT = Integer.MIN_VALUE;
+
+    private final CatalogService catalogService;
+
+    public CheckCatalogVersionOnAppendEntries(CatalogService catalogService) {
+        this.catalogService = catalogService;
+    }
+
+    @Override
+    public @Nullable Message intercept(RaftServerService service, AppendEntriesRequest request, RpcRequestClosure done) {
+        if (request.entriesList() == null || request.data() == null) {
+            return null;
+        }
+
+        Node node = (Node) service;
+
+        ByteBuffer allData = request.data().asReadOnlyByteBuffer();
+        int offset = 0;
+
+        for (RaftOutter.EntryMeta entry : request.entriesList()) {
+            int requiredCatalogVersion = readRequiredCatalogVersionForMeta(allData, entry, node.getOptions().requiredCommandsMarshaller());
+
+            if (requiredCatalogVersion != NO_VERSION_REQUIREMENT && !isMetadataAvailableFor(requiredCatalogVersion)) {
+                LOG.warn("Metadata not yet available, group {}, required level {}.", request.groupId(), requiredCatalogVersion);
+                return RaftRpcFactory.DEFAULT //
+                    .newResponse(node.getRaftOptions().getRaftMessagesFactory(), RaftError.EBUSY,
+                            "Metadata not yet available, group '%s', required level %d.", request.groupId(), requiredCatalogVersion);
+            }
+
+            offset += (int) entry.dataLen();
+            allData.position(offset);
+        }
+
+        return null;
+    }
+
+    private static int readRequiredCatalogVersionForMeta(ByteBuffer allData, final EntryMeta entry, Marshaller commandsMarshaller) {
+        if (entry.type() != EntryType.ENTRY_TYPE_DATA) {
+            return NO_VERSION_REQUIREMENT;
+        }
+
+        if (!(commandsMarshaller instanceof PartitionCommandsMarshaller)) {
+            return NO_VERSION_REQUIREMENT;
+        }
+
+        PartitionCommandsMarshaller partitionCommandsMarshaller = (PartitionCommandsMarshaller) commandsMarshaller;
+
+        long dataLen = entry.dataLen();
+        if (dataLen > 0) {
+            return partitionCommandsMarshaller.readRequiredCatalogVersion(allData);
+        }
+
+        return NO_VERSION_REQUIREMENT;
+    }
+
+    private boolean isMetadataAvailableFor(int requiredCatalogVersion) {
+        return requiredCatalogVersion <= catalogService.latestCatalogVersion();
```

Comment:
I'm a bit confused with the javadoc of `CatalogService#latestCatalogVersion`

> <p>NOTE: This method should only be used at the start of components that may be removed or moved in the future. 

which is not true anymore